File: Dished/View Controllers/DAMenuViewController.m
//
//  DAMenuViewController.m
//  Dished
//
//  Created by Ryan Khalili on 10/18/14.
//  Copyright (c) 2014 Dished. All rights reserved.
//

#import "DAMenuViewController.h"
#import "UIImageView+WebCache.h"
#import "DASettingsViewController.h"
#import "DAContainerViewController.h"
#import "DAUserProfileViewController.h"
#import "DAUserManager.h"
#import "UserVoice.h"


@interface DAMenuViewController()

@property (nonatomic) BOOL initialViewAppear;

@end


@implementation DAMenuViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.initialViewAppear = YES;
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(setupUserInfo) name:kUserProfileUpdatedNotification object:nil];
    
    UIImageView *backgroundImage = [[UIImageView alloc] initWithFrame:self.tableView.frame];
    backgroundImage.image = [UIImage imageNamed:@"menu_background"];
    self.tableView.backgroundView = backgroundImage;
    
    self.userImageView.layer.borderColor   = [UIColor whiteColor].CGColor;
    self.userImageView.layer.borderWidth   = 2;
    self.userImageView.layer.cornerRadius  = self.userImageView.frame.size.width / 2;
    self.userImageView.layer.masksToBounds = YES;
    
    UITapGestureRecognizer *tapGesture1 = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(editProfile)];
    self.userImageView.userInteractionEnabled = YES;
    [self.userImageView addGestureRecognizer:tapGesture1];
    
    UITapGestureRecognizer *tapGesture2 = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(editProfile)];
    self.usernameLabel.userInteractionEnabled = YES;
    [self.usernameLabel addGestureRecognizer:tapGesture2];
    
    [self setupUserInfo];
}

- (void)setupUserInfo
{
    NSURL *userImageURL = [NSURL URLWithString:[DAUserManager sharedManager].img_thumb];
    [self.userImageView sd_setImageWithURL:userImageURL placeholderImage:[UIImage imageNamed:@"profile_image"]];
    
    self.usernameLabel.text = [NSString stringWithFormat:@"@%@", [DAUserManager sharedManager].username];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.navigationController setNavigationBarHidden:YES animated:animated];
    
    if( !self.initialViewAppear && [self.containerViewController menuShowing] )
    {
        [self.containerViewController slideOutMenu];
    }
    
    self.initialViewAppear = NO;
    
    NSIndexPath *selectedRowIndexPath = [self.tableView indexPathForSelectedRow];
    [self.tableView deselectRowAtIndexPath:selectedRowIndexPath animated:YES];
    
    [[self transitionCoordinator] notifyWhenInteractionEndsUsingBlock:^( id<UIViewControllerTransitionCoordinatorContext> context )
    {
        if( [context isCancelled] )
        {
            [self.tableView selectRowAtIndexPath:selectedRowIndexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
            [self.containerViewController moveToMenu];
        }
    }];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [self.containerViewController moveToMenu];
    
    switch( indexPath.row )
    {
        case 0: [self goToInviteFriends]; break;
        case 1: [self goToSettings]; break;
        case 2: [self goToFAQ]; break;
        case 3: [self showTermsOfUse]; break;
        case 4: [self showPrivacyPolicy]; break;
        case 5: [UserVoice presentUserVoiceContactUsFormForParentViewController:self]; break;
    }
}
             
- (void)goToFAQ
{
    [self goToDocumentViewWithName:@"FAQ"];
}

- (void)showTermsOfUse
{
    [self goToDocumentViewWithName:kTermsAndConditions];
}

- (void)showPrivacyPolicy
{
    [self goToDocumentViewWithName:kPrivacyPolicy];
}

- (void)goToDocumentViewWithName:(NSString *)documentName
{
    NSString *filePath = [[NSBundle mainBundle] pathForResource:documentName ofType:@"html"];
    DADocViewController *docVC = [[DADocViewController alloc] initWithFilePath:filePath title:documentName];
    [self.navigationController pushViewController:docVC animated:YES];
}

- (void)goToInviteFriends
{
    DAGetSocialViewController *getSocialVC = [[DAGetSocialViewController alloc] init];
    [self.navigationController pushViewController:getSocialVC animated:YES];
}

- (void)goToSettings
{
    [self pushSettingsView];
}

- (IBAction)editProfile
{
    [self.containerViewController moveToMenu];
    
    [self pushUserProfileWithUsername:[DAUserManager sharedManager].username];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [self.navigationController setNavigationBarHidden:NO animated:animated];
}

@end

File: Dished/View Controllers/DAMenuViewController.swift
//
//  DAMenuViewController2.swift
//  Dished
//
//  Created by Ryan Khalili on 4/17/15.
//  Copyright (c) 2015 Dished. All rights reserved.
//

import UIKit

class DAMenuViewController: DAViewController, UITableViewDelegate, UITableViewDataSource, DAMenuTableViewDelegate {

    var menuView = DAMenuView()
    
    let cellIdentifier = "menuCell"
    let rowTitles = [
        "Find Friends",
        "Settings",
        "FAQs",
        "Terms & Conditions",
        "Privacy Policy",
        "Help"
    ]
    
    private var initialViewAppear = true
    
    override func viewDidLoad() {
        super.viewDidLoad()

        menuView.tableView.delegate = self
        menuView.tableView.dataSource = self
        menuView.tableView.registerClass(UITableViewCell.self, forCellReuseIdentifier: cellIdentifier)
        
        setupUserInfo()
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        navigationController?.setNavigationBarHidden(true, animated: animated)
        
        if !initialViewAppear && containerViewController().menuShowing() {
            containerViewController().slideOutMenu()
        }
        
        initialViewAppear = false
        
        menuView.tableView.deselectSelectedIndexPath()
        
        transitionCoordinator()?.notifyWhenInteractionEndsUsingBlock({
            handler in
            
            if handler.isCancelled() {
                let selectedIndexPath = self.menuView.tableView.indexPathForSelectedRow()
                self.menuView.tableView.selectRowAtIndexPath(selectedIndexPath, animated: false, scrollPosition: UITableViewScrollPosition.None)
                self.containerViewController().moveToMenu()
            }
        })
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        navigationController?.setNavigationBarHidden(false, animated: animated)
    }
    
    func setupUserInfo() {
        let imageURL = NSURL(string: DAUserManager.sharedManager().img_thumb)
        menuView.tableView.userImageView.sd_setImageWithURL(imageURL, placeholderImage: UIImage(named: "profile_image"))
        
        if DAUserManager.sharedManager().username != nil {
            let username = "@\(DAUserManager.sharedManager().username)"
            menuView.tableView.usernameButton.setTitle(username, forState: UIControlState.Normal)
        }
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(cellIdentifier) as! UITableViewCell
        
        cell.textLabel?.text = rowTitles[indexPath.row]
        cell.textLabel?.textColor = UIColor.whiteColor()
        cell.backgroundColor = UIColor.clearColor()
        cell.accessoryType = UITableViewCellAccessoryType.DisclosureIndicator
        
        return cell
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return rowTitles.count
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        containerViewController().moveToMenu()
        
        switch(indexPath.row) {
            case 0: pushGetSocialView()
            case 1: pushSettingsView()
            case 2: goToDocumentViewWithName("FAQ")
            case 3: goToDocumentViewWithName(kTermsAndConditions)
            case 4: goToDocumentViewWithName(kPrivacyPolicy)
            case 5: UserVoice.presentUserVoiceContactUsFormForParentViewController(self)
            default: menuView.tableView.deselectSelectedIndexPath()
        }
    }
    
    private func pushGetSocialView() {
        let getSocialViewController = DAGetSocialViewController()
        navigationController?.pushViewController(getSocialViewController, animated: true)
    }
    
    private func goToDocumentViewWithName(name: String) {
        if let filePath = NSBundle.mainBundle().pathForResource(name, ofType: "html") {
            let documentViewController = DADocViewController(filePath: filePath, title: name)
            navigationController?.pushViewController(documentViewController, animated: true)
        }
    }
    
    func goToUserProfile() {
        containerViewController().moveToMenu()
        pushUserProfileWithUsername(DAUserManager.sharedManager().username)
    }
    
    func userImageTappedOnMenuTableView(menuTableView: DAMenuTableView) {
        goToUserProfile()
    }
    
    override func loadView() {
        view = menuView
        
        menuView.tableView.usernameButton.addTarget(self, action: "goToUserProfile", forControlEvents: UIControlEvents.TouchUpInside)
        menuView.tableView.viewProfileButton.addTarget(self, action: "goToUserProfile", forControlEvents: UIControlEvents.TouchUpInside)
        menuView.tableView.headerDelegate = self
    }
}